enhancements/cert-manager-proxy/http01_challenge_proxy.md

```diff
@@ -0,0 +1,404 @@
+
+- **Proxy Code**: [GitHub Repository](https://github.com/mvazquezc/cert-mgr-http01-proxy/tree/main)
+- **Deployment Manifest**: [Manifest Link](https://github.com/mvazquezc/cert-mgr-http01-proxy/blob/main/manifests/deploy-in-ocp.yaml)
+
+This enhancement aims to provide a robust solution for managing certificates for the API endpoint, particularly for baremetal customers using non cloud-ready DNS servers. Cloud providers typically offer cloud provider DNS services that integrate directly with cert-manager for DNS01 challenges, but baremetal environments often lack these integrations and rely on HTTP01 challenges instead.
+
+### User Stories
+
+1. **As a cluster administrator**, I want to manage custom certificates for the API endpoint (`api.cluster.example.com`) using an external ACME CA, so that I can ensure secure communication for my cluster's API.
+2. **As a cluster administrator on a baremetal platform**, I want a reliable solution to handle HTTP01 challenges for the API endpoint, even when the endpoint is not managed by OpenShift Ingress, so that I can avoid manual workarounds.
+3. **As a developer**, I want a simple deployment mechanism for the HTTP01 challenge proxy, so that I can easily integrate it into my existing cluster setup.
+
+### Goals
+
+- Provide a reliable mechanism for Cert Manager to complete HTTP01 challenges for the API endpoint (`api.cluster.example.com`) in baremetal environments.
+- Ensure compatibility with various OpenShift topologies, including Standard Clusters, Compact Clusters, and SNO.
+- Low operational complexity to secure communication with a cluster's API endpoint.
+
+### Non-Goals
+
+- This enhancement does not aim to replace or modify the existing OpenShift Ingress functionality.
+- It does not provide support for non-HTTP01 challenge types (e.g., DNS-01).
+- It does not address certificate management for endpoints other than the API endpoint (`api.cluster.example.com`).
+- It does not provide a solution for environments where `nftables` is not supported.
+
+## Proposal
+
+The HTTP01 Challenge Proxy will be implemented as a **core component of the OpenShift payload** and managed by an existing cluster operator, specifically the **cluster-kube-apiserver-operator**, which already manages API endpoint configuration and lifecycle.
+
+**Deployment Architecture:**
+- **Core Integration**: This feature is part of the core OpenShift distribution, not an optional OLM-installable operator
+- **Operator Management**: The cluster-kube-apiserver-operator will deploy and manage the proxy DaemonSet
+- **Conditional Deployment**: The proxy is only deployed when `APIServer.spec.http01ChallengeProxy.mode` is set to `DefaultDeployment` or `CustomDeployment`
+- **Platform Targeting**: Limited to baremetal platforms where HTTP01 challenges are needed for API certificates
```

Comment:
> https://github.com/openshift/enhancements/pull/1773#discussion_r2215490405
> https://github.com/openshift/enhancements/pull/1773#pullrequestreview-3176944091 

Regarding the use of this feature on non-baremetal platforms, I understand that the motivation and focus are derived from baremetal, but I still believe we shouldn't prevent non-baremetal users from enabling it with additional validations.

As cert-manager users, while major cloud providers already offer DNS provider support, enabling this feature would provide more options. Users could evaluate and adopt it based on their needs, and this would also represent additional flexibility that OpenShift can offer. What do you think? @mvazquezc @JoelSpeed

Reply:
Taking AWS as an example, does the API server load balancer come configured today to forward traffic on port 80 to the control plane nodes? I suspect it does not.

If it does not, enabling this feature on that platform would achieve nothing? The proxy could be exposed and handle traffic incoming on port 80, but there would still be no way an ACME challenge could hit the proxy

So if my assumption there is correct, what would be the point in allowing someone to configure this on AWS?